### DISCUSSION TITLE
non ASCII characters in metadata

### DISCUSSION THREAD

**Discussion by ruspaul013** [minio/minio#19576] [Q&A] [answered]
Hello, 

I was trying to upload a file using non ASCII characters in metadata and I receive `<Error><Code>SignatureDoesNotMatch</Code><Message>The request signature we calculated does not match the signature you provided. Check your key and signing method.</Message>`.

Here is an example of test file:
[minio-example.txt](https://github.com/minio/minio/files/15062912/minio-example.txt)

It will have support in the future for this? I saw [here](https://github.com/minio/minio-py/issues/764) that `Each key-value pair must conform to US-ASCII when you are using REST`, but I just want to double check.

**Comment by klauspost** [comment]
[Tags](https://docs.aws.amazon.com/AmazonS3/latest/userguide/tagging-managing.html) support unicode. 

Alternative use URL encoding for your values.

I will see if there are changes needed in the python sdk or on the MinIO server.

**Reply by balamurugana** [reply]
@klauspost 

1. The example file is a PHP code. I don't know whether we have minio-php.

2. minio-py works with unicode strings like for `metadata` user has to `helpers.quote()` values otherwise `ValueError: unsupported metadata value Hello, 世界; only US-ASCII encoded characters are supported` is thrown. For `tags`, minio server fails with `minio.error.S3Error: S3 operation failed; code: InvalidTag, message: The TagValue you have provided is invalid, resource: /my-bucket/my-object, request_id: 17C8A1C904024C36, host_id: 3e996b2f640d7e065d3a5c4e39a5538cefb82e3e77771990265e4698d8681eac, bucket_name: my-bucket, object_name: my-object` for encoded tag value `X-Amz-Tagging: User=Hello%2C%20%E4%B8%96%E7%95%8C`

**Reply by ruspaul013** [reply]
Sorry, I forgot to mention that I am using aws-sdk for php😅. To be honest, I did not tried the clients provided by minio. Mostly we use the php client. I also saw that S3 gives the same results.

**Comment by stevapple** [comment]
Currently MinIO doesn't support this feature. S3 documentation is ambiguous (and changing) about how it handles non-ASCII metadata, but its behavior is, in some way, replicable.

I wonder if MinIO is interested in replicating such behavior, which is barely discoverable and even unsupported by some popular client SDKs.

**Comment by harshavardhana** [comment, marked as answer]
AWS S3 does not support unicode characters however they do in their other services, its not documented you have to manually test it to figure this out and we did so we have kept the relevant limits.
